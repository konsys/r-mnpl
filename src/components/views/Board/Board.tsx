import React from "react";
import { Field, BoardField } from "../Field/Field";
import nanoid from "nanoid";

interface Prop {
  fields: BoardField[];
}
export const Board = (prop: Prop) => (
  <div id="ui-fields" className="table-body-board-fields">
    {prop.fields.map((field) => {
      return (
        <Field
          key={nanoid(4)}
          fieldPosition={field.fieldPosition}
          price={field.price}
          imgSrc={field.imgSrc}
          mnplGroup={field.mnplGroup}
          mnplLine={field.mnplLine}
          name={field.name}
          mnplSpecial={field.mnplSpecial}
          mnplCorner={field.mnplCorner}
          isJail={field.isJail}
          status={field.status}
        />
      );
    })}
  </div>
);
