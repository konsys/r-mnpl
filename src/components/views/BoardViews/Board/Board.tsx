import {
  FieldType,
  IField,
  IFieldModalPosition,
} from "../../../../types/types";
import React, { useEffect } from "react";
import {
  addFieldToContract,
  contract$,
} from "../../../../stores/Board/ContractStore";
import {
  closeFieldActionEvent,
  fieldActionStore,
  fieldsStore,
  setFieldActionEffect,
} from "../../../../stores/Board/FieldsStore";

import { BOARD_PARAMS } from "../../../../params/boardParams";
import { Field } from "../Field/Field";
import { FieldActions } from "../Field/FieldActions/FieldAction";
import { useStore } from "effector-react";
import { user$ } from "../../../../stores/Game/UserStore";

export const Board = () => {
  const { fields } = useStore(fieldsStore);
  const fieldActionId = useStore(fieldActionStore);
  const contract = useStore(contract$);
  const user = useStore(user$);

  const getFieldActionPosition = (field: IField): IFieldModalPosition => {
    switch (field.fieldLine) {
      case 0:
        return {
          top: 105,
          left: 25 + (field.fieldPosition - 1) * BOARD_PARAMS.FIELD_WIDTH,
        };
      case 1:
        return field.fieldPosition < 16
          ? {
              top: 105 + (field.fieldPosition - 11) * BOARD_PARAMS.FIELD_WIDTH,
              left: 380,
            }
          : {
              top: (field.fieldPosition - 13) * BOARD_PARAMS.FIELD_WIDTH,
              left: 380,
            };
      case 2:
        return {
          top: 240,
          left: 450 - (field.fieldPosition - 21) * BOARD_PARAMS.FIELD_WIDTH,
        };

      case 3:
        return field.fieldPosition > 34
          ? {
              top: 105 + (39 - field.fieldPosition) * BOARD_PARAMS.FIELD_WIDTH,
              left: 105,
            }
          : {
              top: (39 - field.fieldPosition) * BOARD_PARAMS.FIELD_WIDTH - 110,
              left: 105,
            };

      default:
        return { top: 240, left: 105 };
    }
  };

  const closeFieldAction = (event: any) => {
    (!event.target && !event.target.id && closeFieldActionEvent()) ||
      (event.target.id &&
        !(event.target.id.indexOf("field") > -1) &&
        closeFieldActionEvent());
  };

  useEffect(() => {
    document.addEventListener("click", closeFieldAction, false);
    return () => {
      document.removeEventListener("click", closeFieldAction, false);
    };
  }, []);

  const onClick = (f: IField) => {
    if (
      f.type === FieldType.AUTO ||
      f.type === FieldType.COMPANY ||
      f.type === FieldType.IT
    ) {
      if (contract && user && contract.fromUserId === user.userId) {
        addFieldToContract({
          fromUserId: user.userId,
          toUserId: contract.toUserId,
          field: f,
        });
      } else {
        setFieldActionEffect(f.fieldId || 0);
      }
    } else {
      closeFieldActionEvent();
    }

    return false;
  };

  return (
    <>
      <div id="ui-fields" className="table-body-board-fields">
        {fields &&
          fields.map((f, index) => (
            <Field
              key={(f && f.fieldId) || index}
              {...f}
              onClick={() => onClick(f)}
            />
          ))}
        {fields &&
          fields.map((field, index) => {
            return (
              field.fieldGroupName && (
                <FieldActions
                  key={(field && field.fieldId) || index}
                  {...field}
                  position={getFieldActionPosition(field)}
                  isActive={field.fieldId === fieldActionId}
                />
              )
            );
          })}
      </div>
    </>
  );
};
