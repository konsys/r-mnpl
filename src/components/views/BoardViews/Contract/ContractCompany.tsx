import { IField } from "../../../../types/types";
import React from "react";

export const ContractCompany = ({ field }: { field: IField }) => {
  const price = `${field.price?.startPrice}k` || "0k";
  return (
    <>
      {field.fieldId && (
        <div className="_one _field _deletable">
          <div
            className="_image"
            style={{
              backgroundImage: `url("${field.imgSrc}")`,
            }}
          ></div>
          <div className="_info">
            <div className="_title">{field.name}</div>
            <div className="_subtitle">{price}</div>
          </div>
        </div>
      )}
    </>
  );
};
