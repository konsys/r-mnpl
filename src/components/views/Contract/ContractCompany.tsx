import { IField } from "../../../types/types";
import React from "react";

export const ContractCompany = ({ field }: { field: IField }) => {
  return (
    <>
      {field.fieldId && (
        <div className="_one _field _deletable">
          <div
            className="_image"
            style={{
              backgroundImage:
                'url("https://m1.dogecdn.wtf/fields/brands/2_clothing/lacoste.svg")',
            }}
          ></div>
          <div className="_info">
            <div className="_title">{field.name}</div>
            <div className="_subtitle">{field.price?.startPrice}k</div>
          </div>
        </div>
      )}
    </>
  );
};
