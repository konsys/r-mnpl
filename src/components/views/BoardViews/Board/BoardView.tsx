import {
  FieldType,
  IContract,
  IField,
  IFieldModalPosition,
  IUser,
} from "../../../../types/types";
import React, { useEffect } from "react";

import { BOARD_PARAMS } from "../../../../params/boardParams";
import { Field } from "../Field/Field";
import { FieldActions } from "../Field/FieldActions/FieldAction";
import { IOpenContractModal } from "../../../../stores/Board/ContractStore";

export const BoardView = ({
  fields,
  fieldActionId,
  contract,
  user,
  setFieldActionEffect,
  addFieldToContract,
  closeFieldActionEvent,
}: {
  fields: IField[];
  fieldActionId: number;
  contract: IContract;
  user: IUser | null;
  setFieldActionEffect: (params: any) => any;
  addFieldToContract: (params: IOpenContractModal) => any;
  closeFieldActionEvent: any;
}) => {
  const getFieldActionPosition = (field: IField): IFieldModalPosition => {
    switch (field.fieldLine) {
      case 0:
        return {
          top: 105,
          left: 25 + (field.fieldPosition - 1) * BOARD_PARAMS.FIELD_WIDTH,
        };
      case 1:
        return field.fieldPosition < 16
          ? {
              top: 105 + (field.fieldPosition - 11) * BOARD_PARAMS.FIELD_WIDTH,
              left: 380,
            }
          : {
              top: (field.fieldPosition - 13) * BOARD_PARAMS.FIELD_WIDTH,
              left: 380,
            };
      case 2:
        return {
          top: 240,
          left: 450 - (field.fieldPosition - 21) * BOARD_PARAMS.FIELD_WIDTH,
        };

      case 3:
        return field.fieldPosition > 34
          ? {
              top: 105 + (39 - field.fieldPosition) * BOARD_PARAMS.FIELD_WIDTH,
              left: 105,
            }
          : {
              top: (39 - field.fieldPosition) * BOARD_PARAMS.FIELD_WIDTH - 110,
              left: 105,
            };

      default:
        return { top: 240, left: 105 };
    }
  };

  const closeFieldAction = (event: any) => {
    (!event.target && !event.target.id && closeFieldActionEvent()) ||
      (event.target.id &&
        !(event.target.id.indexOf("field") > -1) &&
        closeFieldActionEvent());
  };

  useEffect(() => {
    document.addEventListener("click", closeFieldAction, false);
    return () => {
      document.removeEventListener("click", closeFieldAction, false);
    };
  });

  const onClick = (f: IField) => {
    if (
      f.type === FieldType.AUTO ||
      f.type === FieldType.COMPANY ||
      f.type === FieldType.IT
    ) {
      if (contract && user && contract.fromUserId === user.userId) {
        addFieldToContract({
          fromUserId: user.userId,
          toUserId: contract.toUserId,
          field: f,
        });
      } else {
        setFieldActionEffect(f.fieldId || 0);
      }
    } else {
      closeFieldActionEvent();
    }

    return false;
  };

  return (
    <>
      <div id="ui-fields" className="table-body-board-fields">
        {fields.map((f, index) => (
          <Field
            key={(f && f.fieldId) || index}
            {...f}
            onClick={() => onClick(f)}
          />
        ))}
        {fields.map((field, index) => {
          return (
            field.fieldGroupName && (
              <FieldActions
                key={(field && field.fieldId) || index}
                {...field}
                position={getFieldActionPosition(field)}
                isActive={field.fieldId === fieldActionId}
              />
            )
          );
        })}
      </div>
    </>
  );
};
