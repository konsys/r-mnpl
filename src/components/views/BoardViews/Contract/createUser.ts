import { OutcomeMessageType } from "types/types";
import fetch from "node-fetch";
import { sendBoardAction } from "stores/Board/ActionStore";
import { testContract } from "testMocks/contract";
// TODO remove
export const createUser = async () => {
  const response = await fetch("http://localhost:3001/users", {
    method: "POST",
  });
  const userId = await response.text();
  sendBoardAction({
    action: OutcomeMessageType.OUTCOME_CONTRACT_DECLINE,
    contract: testContract,
  });
  return userId;
};
