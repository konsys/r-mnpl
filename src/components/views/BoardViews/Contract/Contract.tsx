import {
  IContract,
  IField,
  IUser,
  OutcomeMessageType,
} from "../../../../types/types";
import { IDialogStore, showDialog } from "../../../../stores/Board/DialogStore";
import React, { useState } from "react";
import {
  addMoneyToContract,
  closeContractModal,
} from "../../../../stores/Board/ContractStore";
import {
  gameActionFx,
  sendBoardAction,
} from "../../../../stores/Board/ActionStore";

import { ContractCompany } from "./ContractCompany";
import { getField } from "../../../../utils/fields.utils";
import { getPlayer } from "../../../../utils/players.utils";
import { sample } from "lodash";

export enum KeyCode {
  ENTER = 13,
}

export const validateContract = (contract: IContract): null | IDialogStore => {
  if (!contract.fieldIdsFrom.length && !contract.fieldIdsTo.length) {
    return {
      title: "Ошибка",
      message: "В договоре должно быть хотя бы одно поле.",
    };
  } else if (contract.moneyFrom && contract.moneyTo) {
    return {
      title: "Ошибка",
      message: "Наличные в договоре могут быть только с одной стороны.",
    };
  } else if (
    (contract.moneyFrom + contract.fieldFromPrice) / 2 >
      contract.moneyTo + contract.fieldToPrice ||
    (contract.moneyTo + contract.fieldToPrice) / 2 >
      contract.moneyFrom + contract.fieldFromPrice
  ) {
    return {
      title: "Ошибка",
      message:
        "Разница между суммой предлагаемого и запрашиваемого не может превышать 50%.",
    };
  }
  return null;
};

export const Contract = ({
  contract,
  user,
}: {
  contract: IContract;
  user: IUser;
}) => {
  const [activeInput, setActiveInput] = useState<number>(0);
  const [valueFrom, setValueFrom] = useState<string>("");
  const [valueTo, setValueTo] = useState<string>("");

  // gameActionFx.done.watch(() => {
  //   user &&
  //     (user.userId === contract.fromUserId ||
  //       user.userId === contract.toUserId) &&
  //     closeContractModal();
  // });

  sample({
    clock: gameActionFx.done,
    source: gameActionFx,
    fn: () => closeContractModal(),
  });
  const onContractSubmit = () => {
    setActiveInput(0);
    const res = validateContract(contract);
    res
      ? showDialog(res)
      : sendBoardAction({
          action: OutcomeMessageType.OUTCOME_CONTRACT_START,
          contract,
        });
  };

  const onChange = (e: any) => {
    const v = e.target.value.slice(0, 6);
    if (e.target && e.target.name === "from") {
      setValueFrom(v);
    } else if (e.target && e.target.name === "to") {
      setValueTo(v);
    }
  };

  const onKeyPress = (e: any) => {
    if (e.which !== KeyCode.ENTER) return;
    setActiveInput(0);

    const money = Number.parseInt(e.target.value);

    if (e.target.name === "from") {
      addMoneyToContract({
        fromUserId: contract.fromUserId,
        toUserId: contract.toUserId,
        money: !isNaN(money) ? money : 0,
      });
    } else if (e.target.name === "to") {
      addMoneyToContract({
        fromUserId: contract.toUserId,
        toUserId: contract.fromUserId,
        money: !isNaN(money) ? money : 0,
      });
    }

    setValueFrom("");
    setValueTo("");
  };

  const fromUser = getPlayer(contract.fromUserId);
  const toUser = getPlayer(contract.toUserId);

  const contractType =
    user && contract.fromUserId === user.userId ? "from" : "to";

  return (
    <>
      {contract &&
        user &&
        (contract.fromUserId === user.userId ||
          contract.toUserId === user.userId) && (
          <div className="TableContract" style={{}}>
            <div className="TableContract-top">
              <div className="TableContract-top-title">Договор</div>
              <div
                className="TableContract-top-close"
                onClick={() => closeContractModal()}
              ></div>
            </div>
            <div className="TableContract-content">
              <div className="TableContract-content-head">
                <div className="_user_index_0">
                  <div
                    className="_avatar"
                    style={{
                      backgroundImage: `url("${fromUser && fromUser.avatar}")`,
                    }}
                  ></div>
                  <div className="_info">
                    <div className="_nick">Вы</div>
                    <div className="_subtitle">предлагаете</div>
                  </div>
                </div>
                <div className="_user_index_1">
                  <div
                    className="_avatar"
                    style={{
                      backgroundImage: `url("${toUser && toUser.avatar}")`,
                    }}
                  ></div>
                  <div className="_info">
                    <div className="_nick">{`${toUser && toUser.name}`}</div>
                    <div className="_subtitle">отдаёт</div>
                  </div>
                </div>
              </div>
              <div className="TableContract-content-list">
                <div>
                  <div className="scr" scr-active="1">
                    <div
                      className="scr-window"
                      style={{ width: "262px", height: "294px" }}
                    >
                      <div className="scr-content" style={{ width: "212px" }}>
                        <div className="_one _cash _clickable">
                          <div className="_image"></div>
                          <div className="_info">
                            <div
                              className="_title"
                              onClick={() =>
                                setActiveInput(
                                  (fromUser && fromUser.userId) || 0
                                )
                              }
                            >
                              {fromUser && activeInput !== fromUser.userId ? (
                                <>
                                  {contract.moneyFrom}k
                                  <span className="_edit" />
                                </>
                              ) : (
                                <input
                                  type="number"
                                  name="from"
                                  onChange={onChange}
                                  value={valueFrom}
                                  onKeyPress={onKeyPress}
                                />
                              )}
                            </div>
                            <div className="_subtitle">Наличные</div>
                          </div>
                        </div>

                        {!!contract.fieldIdsFrom.length &&
                          contract.fieldIdsFrom.map((fId, k) => {
                            const f = getField(fId);
                            return (
                              <ContractCompany
                                field={f || ({} as IField)}
                                key={k}
                              />
                            );
                          })}
                      </div>
                      <div className="scr-pane" style={{ display: "none" }}>
                        <div
                          className="scr-pane-handler"
                          style={{ transform: "translateY(0px)}" }}
                        ></div>
                      </div>
                    </div>
                  </div>
                </div>
                <div>
                  <div className="scr" scr-active="1">
                    <div
                      className="scr-window"
                      style={{ width: "261px", height: "294px" }}
                    >
                      <div className="scr-content" style={{ width: "211px" }}>
                        <div className="_one _cash _clickable">
                          <div className="_image"></div>
                          <div className="_info">
                            <div
                              className="_title"
                              onClick={() =>
                                setActiveInput((toUser && toUser.userId) || 0)
                              }
                            >
                              {toUser && activeInput !== toUser.userId ? (
                                <>
                                  {contract.moneyTo}k <span className="_edit" />
                                </>
                              ) : (
                                <input
                                  type="number"
                                  name="to"
                                  onChange={onChange}
                                  value={valueTo}
                                  onKeyPress={onKeyPress}
                                />
                              )}
                            </div>
                            <div className="_subtitle">Наличные</div>
                          </div>
                        </div>
                        {!!contract.fieldIdsTo.length &&
                          contract.fieldIdsTo.map((fId, k) => {
                            const f = getField(fId);
                            return (
                              <ContractCompany
                                field={f || ({} as IField)}
                                key={k}
                              />
                            );
                          })}
                      </div>
                      <div className="scr-pane" style={{ display: "none" }}>
                        <div
                          className="scr-pane-handler"
                          style={{ transform: "translateY(0px)" }}
                        ></div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div className="TableContract-content-bottom">
                <div className="_sum">
                  {contract.moneyFrom + contract.fieldFromPrice}
                </div>
                <div className="_text">Общая сумма</div>
                <div className="_sum">
                  {contract.moneyTo + contract.fieldToPrice}
                </div>
              </div>
            </div>
            <div className="TableContract-actions">
              {contractType === "from" ? (
                <div className="_button _accept" onClick={onContractSubmit}>
                  Предложить
                </div>
              ) : (
                <>
                  <div
                    className="_button"
                    onClick={() =>
                      sendBoardAction({
                        action: OutcomeMessageType.OUTCOME_CONTRACT_ACCEPT,
                        contract,
                      })
                    }
                  >
                    Принять
                  </div>
                  <div
                    onClick={() =>
                      sendBoardAction({
                        action: OutcomeMessageType.OUTCOME_CONTRACT_DECLINE,
                        contract,
                      })
                    }
                    className="_button _button_negative"
                  >
                    Отклонить
                  </div>
                </>
              )}

              {/* <div className="_future">Future</div> */}
            </div>
          </div>
        )}
    </>
  );
};
