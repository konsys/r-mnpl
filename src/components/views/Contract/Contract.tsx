import React, { useState } from "react";
import {
  addMoneyToContract,
  closeContractModal,
  contractStore,
  sendContract,
} from "../../../stores/ContractStore";

import { ContractCompany } from "./ContractCompany";
import { IField } from "../../../types/types";
import { getField } from "../../../utils/fields.utils";
import { getPlayer } from "../../../utils/players.utils";
import { showDialog } from "../../../stores/DialogStore";
import { useStore } from "effector-react";
import { userStore } from "../../../stores/UserStore";

export enum KeyCode {
  ENTER = 13,
}

export const Contract = () => {
  const contract = useStore(contractStore);
  const user = useStore(userStore);

  const [activeInput, setActiveInput] = useState<number>(0);
  const [valueFrom, setValueFrom] = useState<string>("");
  const [valueTo, setValueTo] = useState<string>("");

  const onChange = (e: any) => {
    // if (!e.target.value.match(/^[0-9]+$/)) return;
    const v = e.target.value.slice(0, 6);
    if (e.target && e.target.name === "from") {
      setValueFrom(v);
    } else if (e.target && e.target.name === "to") {
      setValueTo(v);
    }
  };

  const onKeyPress = (e: any) => {
    if (e.which !== KeyCode.ENTER) return;
    setActiveInput(0);

    const money = Number.parseInt(e.target.value);

    // if (e.target.value.match(/^[0-9]+$/)) {
    if (e.target.name === "from") {
      addMoneyToContract({
        fromUserId: contract.fromUserId,
        toUserId: contract.toUserId,
        money: !isNaN(money) ? money : 0,
      });
    } else if (e.target.name === "to") {
      addMoneyToContract({
        fromUserId: contract.toUserId,
        toUserId: contract.fromUserId,
        money: !isNaN(money) ? money : 0,
      });
      // setValueTo(e.target.value);
    }
    // }
    setValueFrom("");
    setValueTo("");
  };

  const onSubmit = () => {
    // TODO написать текст ошибок
    if (contract.moneyFrom && contract.moneyTo) {
      showDialog({
        title: "Ошибка",
        message: "Наличные в договоре могут быть только с одной стороны.",
      });
    } else if (!contract.fieldIdsFrom.length && !contract.fieldIdsTo.length) {
      showDialog({
        title: "Ошибка",
        message: "В договоре должно быть хотя бы одно поле.",
      });
    } else if (
      contract.moneyFrom * 2 > contract.moneyTo ||
      contract.moneyTo * 2 > contract.moneyFrom
    ) {
      showDialog({
        title: "Ошибка",
        message:
          "Разница между суммой предлагаемого и запрашиваемого не может превышать 50%.",
      });
    }
    setActiveInput(0);
    sendContract(contract);
  };

  const fromUser = getPlayer(contract.fromUserId);
  const toUser = getPlayer(contract.toUserId);

  return (
    <>
      {contract && contract.fromUserId === user?.userId && (
        <div className="TableContract" style={{}}>
          <div className="TableContract-top">
            <div className="TableContract-top-title">Договор</div>
            <div
              className="TableContract-top-close"
              onClick={() => closeContractModal()}
            ></div>
          </div>
          <div className="TableContract-content">
            <div className="TableContract-content-head">
              <div className="_user_index_0">
                <div
                  className="_avatar"
                  style={{
                    backgroundImage: `url("${fromUser && fromUser.avatar}")`,
                  }}
                ></div>
                <div className="_info">
                  <div className="_nick">Вы</div>
                  <div className="_subtitle">предлагаете</div>
                </div>
              </div>
              <div className="_user_index_1">
                <div
                  className="_avatar"
                  style={{
                    backgroundImage: `url("${toUser && toUser.avatar}")`,
                  }}
                ></div>
                <div className="_info">
                  <div className="_nick">{`${toUser && toUser.name}`}</div>
                  <div className="_subtitle">отдаёт</div>
                </div>
              </div>
            </div>
            <div className="TableContract-content-list">
              <div>
                <div className="scr" scr-active="1">
                  <div
                    className="scr-window"
                    style={{ width: "262px", height: "294px" }}
                  >
                    <div className="scr-content" style={{ width: "212px" }}>
                      <div className="_one _cash _clickable">
                        <div className="_image"></div>
                        <div className="_info">
                          <div
                            className="_title"
                            onClick={() =>
                              setActiveInput((fromUser && fromUser.userId) || 0)
                            }
                          >
                            {fromUser && activeInput !== fromUser.userId ? (
                              <>
                                {contract.moneyFrom}k
                                <span className="_edit" />
                              </>
                            ) : (
                              <input
                                type="number"
                                name="from"
                                onChange={onChange}
                                value={valueFrom}
                                onKeyPress={onKeyPress}
                              />
                            )}
                          </div>
                          <div className="_subtitle">Наличные</div>
                        </div>
                      </div>

                      {!!contract.fieldIdsFrom.length &&
                        contract.fieldIdsFrom.map((fId, k) => {
                          const f = getField(fId);
                          return (
                            <ContractCompany
                              field={f || ({} as IField)}
                              key={k}
                            />
                          );
                        })}
                    </div>
                    <div className="scr-pane" style={{ display: "none" }}>
                      <div
                        className="scr-pane-handler"
                        style={{ transform: "translateY(0px)}" }}
                      ></div>
                    </div>
                  </div>
                </div>
              </div>
              <div>
                <div className="scr" scr-active="1">
                  <div
                    className="scr-window"
                    style={{ width: "261px", height: "294px" }}
                  >
                    <div className="scr-content" style={{ width: "211px" }}>
                      <div className="_one _cash _clickable">
                        <div className="_image"></div>
                        <div className="_info">
                          <div
                            className="_title"
                            onClick={() =>
                              setActiveInput((toUser && toUser.userId) || 0)
                            }
                          >
                            {toUser && activeInput !== toUser.userId ? (
                              <>
                                {contract.moneyTo}k <span className="_edit" />
                              </>
                            ) : (
                              <input
                                type="number"
                                name="to"
                                onChange={onChange}
                                value={valueTo}
                                onKeyPress={onKeyPress}
                              />
                            )}
                          </div>
                          <div className="_subtitle">Наличные</div>
                        </div>
                      </div>
                      {!!contract.fieldIdsTo.length &&
                        contract.fieldIdsTo.map((fId, k) => {
                          const f = getField(fId);
                          return (
                            <ContractCompany
                              field={f || ({} as IField)}
                              key={k}
                            />
                          );
                        })}
                    </div>
                    <div className="scr-pane" style={{ display: "none" }}>
                      <div
                        className="scr-pane-handler"
                        style={{ transform: "translateY(0px)" }}
                      ></div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div className="TableContract-content-bottom">
              <div className="_sum">{contract.moneyFrom}</div>
              <div className="_text">Общая сумма</div>
              <div className="_sum">{contract.moneyTo}</div>
            </div>
          </div>
          <div className="TableContract-actions">
            <div className="_button" onClick={onSubmit}>
              Предложить
            </div>
            {/* <div className="_future">Future</div> */}
          </div>
        </div>
      )}
    </>
  );
};
