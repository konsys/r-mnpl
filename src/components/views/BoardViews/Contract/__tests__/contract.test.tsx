import * as action from "stores/Board/ActionStore";
import * as modal from "stores/Board/ContractStore";

import { Contract, validateContract } from "../Contract";
import { mount, shallow } from "enzyme";

import React from "react";
import renderer from "react-test-renderer";
import { testContract } from "testMocks/contract";
import { testPlayer1 } from "testMocks/user";

jest.mock("stores/Board/ActionStore", () => ({
  sendBoardAction: jest.fn(),
  gameActionFx: { done: { watch: jest.fn() } },
}));

jest.mock("stores/Board/ContractStore", () => ({
  closeContractModal: jest.fn(),
}));

describe("Contract test", () => {
  it("renders correctly", () => {
    shallow(<Contract contract={testContract} user={testPlayer1} />);
  });
  it("matches snapshot", () => {
    const tree = renderer
      .create(<Contract contract={testContract} user={testPlayer1} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("renders accept button", () => {
    expect(
      mount(<Contract contract={testContract} user={testPlayer1} />).find(
        "._accept"
      )
    ).toHaveLength(1);
  });
  it("renders cancel button", () => {
    expect(
      mount(<Contract contract={testContract} user={testPlayer1} />).find(
        ".TableContract-top-close"
      )
    ).toHaveLength(1);
  });
  it("validates right contract with null result", () => {
    const res = validateContract(testContract);
    expect(res).toStrictEqual(null);
  });

  it("validates error with no fields", () => {
    const res = validateContract({
      fieldIdsFrom: [],
      fieldIdsTo: [],
      fieldFromPrice: 0,
      fieldToPrice: 0,
      moneyFrom: 100,
      moneyTo: 0,
      fromUserId: 1,
      toUserId: 2,
    });
    expect(res).toMatchObject({
      title: "Ошибка",
      message: "В договоре должно быть хотя бы одно поле.",
    });
  });

  it("validates money fron both sides", () => {
    const res = validateContract({
      fieldIdsFrom: [1],
      fieldIdsTo: [2],
      fieldFromPrice: 100,
      fieldToPrice: 200,
      moneyFrom: 100,
      moneyTo: 100,
      fromUserId: 1,
      toUserId: 2,
    });
    expect(res).toMatchObject({
      title: "Ошибка",
      message: "Наличные в договоре могут быть только с одной стороны.",
    });
  });

  it("validates money sum", () => {
    const res = validateContract({
      fieldIdsFrom: [1],
      fieldIdsTo: [2],
      fieldFromPrice: 100,
      fieldToPrice: 99,
      moneyFrom: 100,
      moneyTo: 0,
      fromUserId: 1,
      toUserId: 2,
    });
    expect(res).toMatchObject({
      title: "Ошибка",
      message:
        "Разница между суммой предлагаемого и запрашиваемого не может превышать 50%.",
    });
  });

  it("validates money sum", () => {
    const res = validateContract({
      fieldIdsFrom: [1],
      fieldIdsTo: [2],
      fieldFromPrice: 100,
      fieldToPrice: 49,
      moneyFrom: 0,
      moneyTo: 0,
      fromUserId: 1,
      toUserId: 2,
    });
    expect(res).toMatchObject({
      title: "Ошибка",
      message:
        "Разница между суммой предлагаемого и запрашиваемого не может превышать 50%.",
    });
  });

  it("validates error modal", () => {
    const comp = mount(<Contract contract={testContract} user={testPlayer1} />);
    comp.find("._accept").simulate("click");

    expect(action.sendBoardAction).toHaveBeenCalledTimes(1);
    expect(action.gameActionFx.done.watch).toHaveBeenCalledTimes(1);
  });

  // TODO test all Contract view functionality
  // it.skip("validates  modal is closed", () => {
  //   const comp = mount(<Contract contract={testContract} user={testPlayer1} />);
  //   comp.find("._accept").simulate("click");

  //   expect(modal.closeContractModal).toHaveBeenCalledTimes(1);
  // });

  beforeEach(() => jest.clearAllMocks());
});
