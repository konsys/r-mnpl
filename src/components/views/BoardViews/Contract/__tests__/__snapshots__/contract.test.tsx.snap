// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contract test matches snapshot 1`] = `
<div
  className="TableContract"
  style={Object {}}
>
  <div
    className="TableContract-top"
  >
    <div
      className="TableContract-top-title"
    >
      Договор
    </div>
    <div
      className="TableContract-top-close"
      onClick={[Function]}
    />
  </div>
  <div
    className="TableContract-content"
  >
    <div
      className="TableContract-content-head"
    >
      <div
        className="_user_index_0"
      >
        <div
          className="_avatar"
          style={
            Object {
              "backgroundImage": "url(\\"undefined\\")",
            }
          }
        />
        <div
          className="_info"
        >
          <div
            className="_nick"
          >
            Вы
          </div>
          <div
            className="_subtitle"
          >
            предлагаете
          </div>
        </div>
      </div>
      <div
        className="_user_index_1"
      >
        <div
          className="_avatar"
          style={
            Object {
              "backgroundImage": "url(\\"undefined\\")",
            }
          }
        />
        <div
          className="_info"
        >
          <div
            className="_nick"
          >
            undefined
          </div>
          <div
            className="_subtitle"
          >
            отдаёт
          </div>
        </div>
      </div>
    </div>
    <div
      className="TableContract-content-list"
    >
      <div>
        <div
          className="scr"
          scr-active="1"
        >
          <div
            className="scr-window"
            style={
              Object {
                "height": "294px",
                "width": "262px",
              }
            }
          >
            <div
              className="scr-content"
              style={
                Object {
                  "width": "212px",
                }
              }
            >
              <div
                className="_one _cash _clickable"
              >
                <div
                  className="_image"
                />
                <div
                  className="_info"
                >
                  <div
                    className="_title"
                    onClick={[Function]}
                  >
                    <input
                      name="from"
                      onChange={[Function]}
                      onKeyPress={[Function]}
                      type="number"
                      value=""
                    />
                  </div>
                  <div
                    className="_subtitle"
                  >
                    Наличные
                  </div>
                </div>
              </div>
            </div>
            <div
              className="scr-pane"
              style={
                Object {
                  "display": "none",
                }
              }
            >
              <div
                className="scr-pane-handler"
                style={
                  Object {
                    "transform": "translateY(0px)}",
                  }
                }
              />
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          className="scr"
          scr-active="1"
        >
          <div
            className="scr-window"
            style={
              Object {
                "height": "294px",
                "width": "261px",
              }
            }
          >
            <div
              className="scr-content"
              style={
                Object {
                  "width": "211px",
                }
              }
            >
              <div
                className="_one _cash _clickable"
              >
                <div
                  className="_image"
                />
                <div
                  className="_info"
                >
                  <div
                    className="_title"
                    onClick={[Function]}
                  >
                    <input
                      name="to"
                      onChange={[Function]}
                      onKeyPress={[Function]}
                      type="number"
                      value=""
                    />
                  </div>
                  <div
                    className="_subtitle"
                  >
                    Наличные
                  </div>
                </div>
              </div>
            </div>
            <div
              className="scr-pane"
              style={
                Object {
                  "display": "none",
                }
              }
            >
              <div
                className="scr-pane-handler"
                style={
                  Object {
                    "transform": "translateY(0px)",
                  }
                }
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="TableContract-content-bottom"
    >
      <div
        className="_sum"
      >
        240
      </div>
      <div
        className="_text"
      >
        Общая сумма
      </div>
      <div
        className="_sum"
      >
        120
      </div>
    </div>
  </div>
  <div
    className="TableContract-actions"
  >
    <div
      className="_button _accept"
      onClick={[Function]}
    >
      Предложить
    </div>
  </div>
</div>
`;
